datafusion/expr/src/expr_rewriter.rs

```diff
@@ -525,29 +525,55 @@ pub fn coerce_plan_expr_for_schema(
     plan: &LogicalPlan,
     schema: &DFSchema,
 ) -> Result<LogicalPlan> {
-    let new_expr = plan
-        .expressions()
-        .into_iter()
+    match plan {
+        LogicalPlan::Projection(Projection { expr, input, .. }) => {
```

Comment:
```suggestion
        // special case Projection to avoid adding multiple projections
        LogicalPlan::Projection(Projection { expr, input, .. }) => {
```

I didn't understand the need to special case `LogicalPlan::projection` -- I think it would help to explain the rationale 


For anyone who is curious I tried removing this case and got errors like the following:

```



---- sorted_union_with_different_types_and_group_by stdout ----
thread 'sorted_union_with_different_types_and_group_by' panicked at 'assertion failed: `(left == right)`
  left: `"Sort: a ASC NULLS LAST\n  Union\n    Projection: CAST(x.a AS Float64) AS a\n      Projection: x.a\n        Aggregate: groupBy=[[x.a]], aggr=[[]]\n          SubqueryAlias: x\n            Projection: Int64(1) AS a\n              EmptyRelation\n    Projection: x.a\n      Aggregate: groupBy=[[x.a]], aggr=[[]]\n        SubqueryAlias: x\n          Projection: Float64(1.1) AS a\n            EmptyRelation"`,
 right: `"Sort: a ASC NULLS LAST\n  Union\n    Projection: CAST(x.a AS Float64) AS a\n      Aggregate: groupBy=[[x.a]], aggr=[[]]\n        SubqueryAlias: x\n          Projection: Int64(1) AS a\n            EmptyRelation\n    Projection: x.a\n      Aggregate: groupBy=[[x.a]], aggr=[[]]\n        SubqueryAlias: x\n          Projection: Float64(1.1) AS a\n            EmptyRelation"`', datafusion/sql/tests/integration_test.rs:2175:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:65:14
   2: core::panicking::assert_failed_inner
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:246:17
   3: core::panicking::assert_failed
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:203:5
   4: integration_test::quick_test
             at ./tests/integration_test.rs:2175:5
   5: integration_test::sorted_union_with_different_types_and_group_by
             at ./tests/integration_test.rs:1714:5
   6: integration_test::sorted_union_with_different_types_and_group_by::{{closure}}
             at ./tests/integration_test.rs:1700:1
   7: core::ops::function::FnOnce::call_once
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/ops/function.rs:251:5
   8: core::ops::function::FnOnce::call_once
             at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/ops/function.rs:251:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Which are due to another projection having been addeed

Reply:
Yes, it is to avoid adding multiple projections.